lib/Sema/MiscDiagnostics.cpp

```diff
@@ -2587,6 +2587,306 @@ static void diagnoseImplicitSelfUseInClosure(const Expr *E,
   const_cast<Expr *>(E)->walk(DiagnoseWalker(ctx, ACE));
 }
 
+// MARK: -
+
+/// Diagnose cases where binding a strong reference to a weak/unowned capture list
+/// entry implicitly creates a strong capture of the referenced value in an ancestor
+/// escaping closure.
+static void diagnoseImplicitWeakToStrongCapture(const Expr *E,
+                                                const DeclContext *DC) {
+  if (!E || isa<ErrorExpr>(E) || !E->getType())
+    return;
+
+  class Util {
+  public:
+    static ReferenceOwnership getDeclOwnership(const Decl *D) {
+      if (auto attr = D->getAttrs().getAttribute<ReferenceOwnershipAttr>()) {
+        return attr->get();
+      }
+      // Default to strong if unspecified
+      return ReferenceOwnership::Strong;
+    }
+
+    static bool isLessThanStrongOwnership(const ReferenceOwnership ownership) {
+      return isLessStrongThan(ownership, ReferenceOwnership::Strong);
+    }
+
+    static bool isEscapingClosure(const AbstractClosureExpr *ACE) {
+      return !ACE->getType()->isNoEscape();
+    }
+
+    static bool
+    isDeclACaptureListItemOfClosure(const Decl *D,
+                                    const AbstractClosureExpr *ACE) {
+      if (auto VD = dyn_cast<VarDecl>(D))
+        if (auto CLE = VD->getParentCaptureList())
+          if (CLE->getClosureBody() == ACE)
+            return true;
+
+      return false;
+    }
+
+    static std::optional<AbstractClosureExpr *>
+    ancestorEscapingClosureOfDecl(const Decl *D) {
+      DeclContext *DC = D->getDeclContext();
+      while (DC) {
+        if (auto ACE = dyn_cast<AbstractClosureExpr>(DC))
+          if (isEscapingClosure(ACE))
+            return ACE;
+
+        DC = DC->getParent();
+      }
+      return std::nullopt;
+    }
+
+    static bool isDeclDescendantOfEscapingClosure(const Decl *D) {
+      return ancestorEscapingClosureOfDecl(D).has_value();
+    }
+  };
+
+  /// Info regarding weak/unowned capture list items who have simple inits
+  /// binding to a DeclRefExpr with a corresponding Decl that has strong
+  /// ownership (i.e. the capture list item changed ownership from strong
+  /// to weak/unowned).
+  struct WeakToStrongCaptureItemInfo {
+    VarDecl *itemDecl;
+    ValueDecl *itemReferent;
+    DeclRefExpr *itemInitDRE;
+    ReferenceOwnership itemDeclOwnership;
+  };
+
+  class ImplicitWeakToStrongCaptureWalker : public BaseDiagnosticWalker {
+    ASTContext &Ctx;
+
+  public:
+    ImplicitWeakToStrongCaptureWalker(ASTContext &ctx) : Ctx(ctx) {}
```

Comment:
general question: does macro-generated code have to be handled differently when walking the AST? i saw there are some methods in the walker relating to macros but don't really understand the model for how such code should be processed when attempting to produce diagnostics.